server/src/test/java/org/elasticsearch/reservedstate/service/FileSettingsServiceTests.java

```diff
@@ -278,6 +284,58 @@ public void testProcessFileChanges() throws Exception {
         verify(controller, times(1)).process(any(), any(XContentParser.class), eq(ReservedStateVersionCheck.HIGHER_VERSION_ONLY), any());
     }
 
+    @SuppressWarnings("unchecked")
+    public void testInvalidJSON() throws Exception {
+        doAnswer((Answer<Void>) invocation -> {
```

Comment:
BTW, I have a fix for race conditions in `testProcessFileChanges` (https://github.com/elastic/elasticsearch/pull/116309) -- I requested a review from you since some of the same issues apply to `testInvalidJSON`

Reply:
Oh, have I re-introduced a race? 😱

Reply:
Not really re-introduced, it's just that the new test shares the same issues as `testProcessFileChanges` which I merged a fix for this morning (the issues too long to re-capitulate but with the full description in https://github.com/elastic/elasticsearch/pull/116309)

Looking closer at the test though, I think it might be all good -- except that you might need to add a `REPLACE_EXISTING` flag to `writeTestFile` (my PR updated that one this morning)

Reply:
I think I understand... I sort of undid [this fix](https://github.com/elastic/elasticsearch/pull/116309/files#diff-72a815310d2c340df8f6daed2c538b65848f854e58de476d4d12dee52fdb6da7R215)?

Reply:
I like your `addFileChangedListener` approach better; seems cleaner than a spy. However, it doesn't seem to get called for the invalid JSON case. (Is that a bug? The file _did_ change.)

Reply:
Alright @n1v0lg I pushed an update. Does this seem better?

Reply:
@prdoyle can you pull upstream changes? `writeTestFile` changed since you pulled

Reply:
Yup believe this should work (with emphasis and should and believe) -- unless the move operation somehow registers as more than one polled event, in which case there might still be a race condition here again... 

I think it's fair to go ahead with this test and see how it fares since theoretically at least it looks correct.  

Could you add a `testWindows` label to the PR to exercise the OS where this can break?